src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs

```diff
@@ -321,6 +328,206 @@ public async Task<GetTransactionModel> GetTransactionAsync(string txid)
+                        var prevTransactionScriptPubkey = prevTransaction.Outputs[txIn.PrevOut.N].ScriptPubKey;
+
+                        var addressBase58 = GetBase58AddressFromScript(prevTransactionScriptPubkey);
+                        if (addressBase58 == null)
+                            continue;
+
+                        addressGroupBase58.Add(addressBase58);
+                    }
+
+                    // Group change addresses with input addresses.
+                    if (addressGroupBase58.Any())
+                    {
+                        foreach (var txOut in tx.Outputs)
+                        {
+                            var txOutAddressBase58 = GetBase58AddressFromScript(txOut.ScriptPubKey);
+                            if (txOutAddressBase58 != null)
+                            {
+                                if (IsChange(wallet, coinType, txOutAddressBase58))
+                                    addressGroupBase58.Add(txOutAddressBase58);
+                            }
+                        }
+
+                        groupings.Add(addressGroupBase58);
+                    }
+                }
+
+                // Group lone addresses by themselves.
+                foreach (var txOut in tx.Outputs)
+                {
+                    if (IsAddressMine(wallet, txOut.ScriptPubKey, coinType))
+                    {
+                        var grouping = new List<BitcoinPubKeyAddress>();
+
+                        var addressBase58 = GetBase58AddressFromScript(txOut.ScriptPubKey);
+                        if (addressBase58 == null)
+                            continue;
+
+                        grouping.Add(addressBase58);
+
+                        groupings.Add(grouping);
+                    }
+                }
+            }
+
+            var uniqueGroupings = new List<List<BitcoinPubKeyAddress>>();
+            var setMap = new Dictionary<BitcoinPubKeyAddress, List<BitcoinPubKeyAddress>>();
+            foreach (var group in groupings)
+            {
+                var hits = new List<List<BitcoinPubKeyAddress>>();
+                foreach (var address in group)
+                {
+                    var mappedEntry = setMap.FirstOrDefault(m => m.Key == address);
+                    if (mappedEntry.Value != null)
+                        hits.Add(mappedEntry.Value);
+
+                    var merged = new List<BitcoinPubKeyAddress>(group);
+                    foreach (var hit in hits)
+                    {
+                        merged.AddRange(hit);
+                        uniqueGroupings.Remove(hit);
+                    }
+
+                    uniqueGroupings.Add(merged);
+
+                    foreach (var item in merged)
+                    {
+                        setMap[item] = merged;
+                    }
+                }
+            }
+
+            return uniqueGroupings;
+        }
+
+        private BitcoinPubKeyAddress GetBase58AddressFromScript(Script script)
```

Comment:
Maybe consider using `IScriptAddressReader.GetAddressFromScriptPubKey()` for this? It should be available via the wallet manager.

Reply:
I've implemented this now :)